### PR TITLE
Implement component to publish metrics to Azure monitor

### PR DESCRIPTION
**Pull request by yetamsft**

### CODE REVIEW

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+$schema: http://azureml/sdk-2-0/SparkComponent.json
+type: spark
+
+name: model_monitor_metric_outputter
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml:4]: should this be model_monitor_metric_publisher

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  signal_metrics:
+    type: mltable
+    mode: direct
+  monitor_name:
+    type: string
+  signal_name:
+    type: string
+  location:
+    type: string
+  workspace_resource_id:
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml:24]: do you have that information from within the component?

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+inputs:
+  signal_metrics:
+    type: mltable
+    mode: direct
+  monitor_name:
+    type: string
+  signal_name:
+    type: string
+  location:
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml:22]: Is there a way we don't need to be location aware for publishing custom metrics?

File: assets/model_monitoring/components/src/model_monitor_metric_publisher/azure_monitor_metric_publisher.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import datetime
+import json
+import requests
+
+from pyspark.sql import Row
+from typing import List
+
+from shared_utilities.log_utils import log_utils
+
+def publish_metric(
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/src/model_monitor_metric_publisher/azure_monitor_metric_publisher.py:10]: We should probably only publish metrics from the metric publisher component - the outputter only creates the monitoring metric files right?

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml (hunk 2 above)
Comment (yetamsft) [reply assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml:24]: I did find the `workspace_resource_id` from the runtime env vars (`os.environ['AZUREML_WORKSPACE_SCOPE']`), assuming the workspace running the job is always going to be the same where the metrics go. Though I am unsure how reliable this is in future. To me, it is easier to construct (or control) in MoMo.

File: assets/model_monitoring/components/src/model_monitor_metric_publisher/azure_monitor_metric_publisher.py (hunk 4 above)
Comment (yetamsft) [reply assets/model_monitoring/components/src/model_monitor_metric_publisher/azure_monitor_metric_publisher.py:10]: Yes, the outputter changes are not relevant in PR (simple rename), which I can move out to another one for ease to review.

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml (hunk 3 above)
Comment (yetamsft) [reply assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml:22]: Similar to the workspace_resource_id, there is some env var we can grok this infromation from, namely `Env var AZUREML_SERVICE_ENDPOINT: https://eastus2euap.api.azureml.ms`. But again we do not how sure we can depend on them. Thoughts?

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml (hunk 1 above)
Comment (yetamsft) [reply assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml]: Fixed.

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml (hunk 2 above)
Comment (yetamsft) [reply assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml]: After a second thought, both of the variables seem to be generally useful for any jobs. For simplicity of the caller, I changed to pull the location and workspace resource id from the env vars, with some assertion to make sure they exists and are valid.

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml (hunk 3 above)
Comment (yetamsft) [reply assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml]: Same to the comment above. Updated to extract from the env var.

File: assets/model_monitoring/components/src/model_monitor_metric_publisher/azure_monitor_metric_publisher.py (hunk 4 above)
Comment (yetamsft) [reply assets/model_monitoring/components/src/model_monitor_metric_publisher/azure_monitor_metric_publisher.py:11]: Moved the irrelevant changes out.

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py [hunk 5]
```diff
@@ -20,13 +20,6 @@
 from shared_utilities.constants import MDC_CORRELATION_ID_COLUMN, MDC_DATA_COLUMN
 
 
-def _format_date_string(format: str, date_to_format: str) -> datetime:
```
Comment (chupadhy) [inline assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:23]: Nice moving it to a util

File: assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    arg_parser.add_argument("--monitor_name", type=str)
+    arg_parser.add_argument("--signal_name", type=str)
+    arg_parser.add_argument("--data_window_start", type=str)
+    arg_parser.add_argument("--data_window_end", type=str)
+    args = arg_parser.parse_args()
+
+    metrics: List[Row] = read_mltable_in_spark(args.signal_metrics).collect()
+
+    try:
+        azureml_service_endpoint = os.environ['AZUREML_SERVICE_ENDPOINT']
```
Comment (chupadhy) [inline assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py:31]: Who sets these envs? Also, it is possible to not have an online endpoint associated with the model monitor no? Then do we expect the env var to be an empty string or something like that?

File: assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        workspace_resource_id = os.environ['AZUREML_WORKSPACE_SCOPE']
+    except KeyError as err:
+        log_error(f"Failed to find a required environment variable, error: {str(err)}.")
+        raise
+
+    location = extract_location_from_aml_endpoint_str(azureml_service_endpoint)
```
Comment (chupadhy) [inline assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py:37]: Is it possible to Get on WorkspaceResourceId to get the location info instead?

Comment (yetamsft) [reply assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py:37]: No, I don't think so. `WorkspaceResourceId` doesn't contain location info, which is like, `/subscriptions/{sub_id}/resourceGroups/{rg_name}/providers/Microsoft.MachineLearningServices/workspaces/{ws_name}`.

File: assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py (hunk 6 above)
Comment (yetamsft) [reply assets/model_monitoring/components/src/model_monitor_metric_publisher/run.py:31]: I think this is set by Execution Service (or AEther), which is also used for syncing AML run metadata etc.
This is not the "online endpoint" resource, but the AML service (cluster) endpoint, which usually looks like "https://eastus2.api.azureml.ms". I believe this is quite stable and unlikely to change. Of course, we can revisit if we see a problem later.

File: assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
+$schema: http://azureml/sdk-2-0/SparkComponent.json
+type: spark
+
+name: model_monitor_metric_publisher
+display_name: Model Monitor - Metric Publisher
```
[the remark this answers is not in the document]
Comment (yetamsft) [reply assets/model_monitoring/components/model_monitor/model_monitor_metric_publisher/spec.yaml]: Not that I am aware of. Done.

**Review by chupadhy — APPROVED** [review]

**Review by mastloui-msft — APPROVED** [review]

**Pull request merged**